zkp/js/lib/util.js

```diff
@@ -15,31 +15,154 @@
 }
 
-function poseidonDecrypt(cipherText, sharedSecret, nonce) {
-  const plainText = [];
-  cipherText.forEach((msg, index) => {
-    const hash = Poseidon.poseidon4([sharedSecret[0], sharedSecret[1], BigInt(nonce), BigInt(index)]);
-    // subtract the hash from the cipherText to get the plainText
-    plainText[index] = addMod([BigInt(msg), -hash], F);
-    while (plainText[index] < 0n) {
-      plainText[index] += F;
+// per the encryption scheme in ../circuits/lib/encrypt.circom,
+// the nonce must not be larger than 2^128
+function newEncryptionNonce() {
+  const hex = randomBytes(16).toString('hex');
+  const nonce = BigInt(`0x${hex}`);
+  return nonce;
+}
+
+const two128 = BigInt('340282366920938463463374607431768211456');
+// Field modulus for BN254
+const F = BigInt('21888242871839275222246405745257275088548364400416034343698204186575808495617');
+
+// Implements the encryption and decryption functions using Poseidon hash
+// as described: https://drive.google.com/file/d/1EVrP3DzoGbmzkRmYnyEDcIQcXVU7GlOd/view
+// The encryption and decryption functions are compatible with the circom implementations,
+// meaning the cipher texts encrypted by the circuit in circuits/lib/encrypt.circom can
+// be decrypted by the poseidonDecrypt function. And vice versa.
+function poseidonEncrypt(msg, key, nonce) {
+  validateInputs(msg, key, nonce);
+
+  // the size of the message array must be a multiple of 3
+  const message = [...msg];
+  while (message.length % 3 > 0) {
+    // pad with zeros if necessary
+    message.push(0n);
+  }
+
+  // Create the initial state
+  // S = (0, kS[0], kS[1], N + l * 2^128)
+  let state = [0n, key[0], key[1], nonce + BigInt(msg.length) * two128];
+
+  const ciphertext = [];
+
+  const n = Math.floor(message.length / 3);
+  for (let i = 0; i < n; i += 1) {
+    // Iterate Poseidon on the state
+    state = poseidon(state, 4);
+
+    // Modify the state for the next round
+    state[1] = addMod(message[i * 3], state[1]);
+    state[2] = addMod(message[i * 3 + 1], state[2]);
+    state[3] = addMod(message[i * 3 + 2], state[3]);
+
+    // Record the three elements of the encrypted message
+    ciphertext.push(state[1]);
+    ciphertext.push(state[2]);
+    ciphertext.push(state[3]);
+  }
+
+  // Iterate Poseidon on the state one last time
+  state = poseidon(state, 4);
+
+  // Record the last ciphertext element
+  ciphertext.push(state[1]);
+
+  return ciphertext;
+}
+
+function poseidonDecrypt(ciphertext, key, nonce, length) {
+  validateInputs(ciphertext, key, nonce, length);
+
+  // Create the initial state
+  // S = (0, kS[0], kS[1], N + l ∗ 2^128).
+  let state = [0n, key[0], key[1], nonce + BigInt(length) * two128];
+
+  const message = [];
+
+  const n = Math.floor(ciphertext.length / 3);
+  for (let i = 0; i < n; i += 1) {
+    // Iterate Poseidon on the state
+    state = poseidon(state, 4);
+
+    // Release three elements of the decrypted message
+    message.push(addMod(ciphertext[i * 3], -state[1]));
```

Comment:
I was confused by the paper which didn't use subtraction for decryption. But I guess it's just the level of details didn't need to express in the papaer

<img width="604" alt="image" src="https://github.com/user-attachments/assets/e2272233-2f89-49d5-8e3d-de65affa4f4f">